### PR TITLE
PR template tweaks

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

I'd noted some recurring themes in our PR submissions that seemed to hold up merging occassionally, so I decided to codify the merge requirements.  They can be changed per PR - this is just a template.

## Affected Issue Numbers

- no issue - just a quick inspired set of edits

## Code Review Notes

I put the "tag specific reviewers" item in case you request say 5 reviewers, only require 2 approvals, but you want one of those reviewers to be a specific reviewer.  I figure most of the time, that item will be deleted.

## Checklist

- [ ] Changes have been added to the *Unreleased* section in the [`changelog.md`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md).
  - Not a part of the codebase
- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- Tests
  - [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -16,11 +33,13 @@ attention to.  E.g. There are significant logic changes in function X.
...
-- [ ] All issue requirements satisfied (or no linked issues)
+- [ ] Changes have been added *(or are not necessary to add)* to the
+      *Unreleased* section of the [`changelog.md`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md).
+- [ ] All issue requirements satisfied *(or no linked issues)*
 - [ ] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
 - [ ] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
 - Tests
-  - [ ] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
+  - [ ] [Tests implemented for new methods *(or no new methods)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
+  - [ ] Tests updated for changed methods *(or no changed methods)*
```
Comment (lparsons) [inline .github/pull_request_template.md:43]: C: These seemed sufficient as one item, perhaps change it to:
```suggestion
  - [ ] [Tests implemented/updated for new/changed methods *(or no new/changed methods)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
```

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -16,11 +33,13 @@ attention to.  E.g. There are significant logic changes in function X.
...
   - [ ] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
-  - [ ] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*
+  - [ ] All example load tests pass (remotely) *(or all failures predate this PR)*
```
Comment (lparsons) [inline .github/pull_request_template.md:45]: C: The loading tests are running automatically via GitHub actions (and create their own green checks/red x's). It seems redundant to me to have a manual checklist item for these.

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
+Briefly describe the changes in this pull request (put details in the
+developer section of the linked issue(s)).
 
-## Affected Issue Numbers
+## Affected Issue/Pull Request Numbers
 
 - Resolves #<issue_number_1>
-- Resolves #<issue_number_2>
+- Partially addresses #<issue_number_2>
+- Depends on parent #<pull_request_number>
```
Comment (lparsons) [inline .github/pull_request_template.md:11]: Nice to have some examples to refer to issues that aren't being closed.

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -1,13 +1,30 @@
...
+- [x] Minimum approvals: 2
+- [x] Changes requested: 0
+- [ ] Specific approvals: @\<specific reviewers who MUST approve>
+- [ ] Child pull requests are merged:
+  - #<pull_request_number_1>
+  - #<pull_request_number_2>
+- Basic requirements
+  - [x] Linting passes
+  - [x] Tests pass
+  - [x] Conflicts resolved
```
Comment (lparsons) [inline .github/pull_request_template.md:27]: I like the idea of refining the process a bit, esp. as we move to a maintenance phase for this project. However, this sort of process seems like something we should discuss before implementing. Perhaps at our next group meeting?

**Review by lparsons — COMMENTED** [review]
Overall these look like nice changes. While we're at it though, I'd like to review the process and rules with the group before we implement them in the template.

File: .github/pull_request_template.md (hunk 4 above)
Comment (hepcat72) [reply .github/pull_request_template.md:27]: Sure.

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -16,11 +33,13 @@ attention to.  E.g. There are significant logic changes in function X.
 
 ## Checklist
 
-- [ ] Changes have been added to the *Unreleased* section in the [`changelog.md`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md).
-- [ ] All issue requirements satisfied (or no linked issues)
+- [ ] Changes have been added *(or are not necessary to add)* to the
+      *Unreleased* section of the [`changelog.md`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md).
+- [ ] All issue requirements satisfied *(or no linked issues)*
 - [ ] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
 - [ ] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply .github/pull_request_template.md:40]: You would think.  There have been cases where linting errors are unrelated to the changes in the PR.  If someone else is dealing with fixing those linting issues and everything passes locally, I think, to enable concurrency of separate efforts, marking that all linting passes and that you intend to merge despite linting errors shown on the PR clarifies the "reviewability" of the PR and the intentions of the author.

But I'm not married to these items.  I just find them helpful/clarifying.

File: .github/pull_request_template.md (hunk 1 above)
Comment (hepcat72) [reply .github/pull_request_template.md:43]: I suppose.  I was just trying to reduce my need for adding clarifying notes under these items to specify what I've done.  I will often update tests for changed methods and indicate that I didn't add a test for some new method because it's trivial or something.

- [ ] Merge the new/changed method tests items

File: .github/pull_request_template.md (hunk 2 above)
Comment (hepcat72) [reply .github/pull_request_template.md:45]: I guess I'm the only one that adds in details as to why an item is unchecked or how only part of it is done, or why the check applies to the PR even though the CI test has a red checkmark.  Edge cases, for sure.

- [ ] Remove checlist items that overlap the CI tests

Comment (lparsons) [reply .github/pull_request_template.md:45]: I think it's important to explain those edge cases. It makes sense to me that no explanation is needed if there are green checks (things all passing). If there are red X's, then we shouldn't merge without some explanation. Basically, that would be an exception to the rule that all tests must be passing.

Comment (hepcat72) [reply .github/pull_request_template.md]: Added a directive to explain unchecked items.

**Review by lparsons — COMMENTED** [review]
I like the idea of clarifying and making some of the requirements and checks explicit, thanks for taking this on @hepcat72.

However, I feel like this sort of process is something that requires buy-in from all of the developers. I'd propose that we create a process document that outlines our general principals and rules. We can add some of the likely missed things as checklist items in the PR template as you've done here.

A process document could be more verbose and include the _why_ for the parts of the process. It will also give us an opportunity to develop the process as a team. I've made this an agenda item for our next TraceBase meeting.

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -1,26 +1,50 @@
...
+  - Specific approvals required
+    <!-- Approvals from the contributors you select are required regardless of
+    minimum approvals.  Check or delete as desired. -->
+    - [ ] @cbartman1
+    - [ ] @hepcat72
+    - [ ] @jcmatese
+    - [ ] @fkang-pu
+    - [ ] @lparsons
+    - [ ] @mneinast
+    - [ ] @narzouni
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply .github/pull_request_template.md:39]: Huh.  I swear I changed this the other day to remove the `@` symbols.  I also made other changes.  I wonder if they never got in either...

Oh wait.  Different project's template.

So there is a difference between "request reviews" and requiring specific reviews here that has caused delays in PR merging and ambiguity as to whether a reviewer's review is required.  This section solves that problem.

Take this example.  If there is a minimum number of reviews required for merge of 2, and 3 reviewers are requested, that allows for some flexibility in the review process.  Reviewers who don't have the time, can put it off and feel like maybe they're not causing a delay.  However, if one reviewer has specific domain knowledge that the author wishes to leverage, that's where this section comes in handy.  The minimum number or review approvals could be met, but merge cannot happen until the reviewers specified in this section have done a review.

This has been a particular frustration of mine in deciding whether to wait on a third review or not.  I think it provides clarity as to what's being waited on that is otherwise lacking without this section.

[the remark this answers is not in the document]
Comment (hepcat72) [reply .github/pull_request_template.md:39]: - [ ] TODO: Remove `@` tags

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -1,26 +1,50 @@
...
+- Associated issue/pull request requirements:
+  - [x] All resolved issue requirements satisfied
+  - [x] All dependent pull requests are merged *(or none)*
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply .github/pull_request_template.md:42]: I disagree with your premise.  Specific efforts often depend on changes in other branches.  They build on one another.  If the PR is submitted based off main, the changes displayed will include the changes on the parent branch, which means that reviewers would have to sift through already-reviewed changes, which would be like re-inventing the wheel.  Those already-reviewed changes can be obfuscated by basing the PR off the parent branch, instead of main.  This causes only the differences with the parent branch to be displayed.

It's my opinion that these PRs, based on parent branches, are ready for review at any time.  There is of course, a small risk that changes to the parent branch could trigger significant changes to the dependent branch, but I would point out that we just reviewed about 2 dozen such hierarchical branches and there were virtually no changes - and zero significant changes.  And that evidence would suggest that reviewing these PRs earlier would not have changed anything.

But it's important for the reviewer to know that a particular PR is based on changes in another branch.  Reviewing the PRs in order from the root-most branch to the leaf-most branch would helpo the reviewer to become familiar with the underlying code, so despite our difference of opinion on whether PR's should be based on main versus its parent branch, this information is useful to the reviewer.

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -1,26 +1,50 @@
...
+- Basic requirements
+  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
+  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
+  - [x] All conflicts resolved
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply .github/pull_request_template.md:46]: These are not provided for the author to think about each time.  They are here for the author to explicitly acknowledge (i.e. uncheck) that they are aware of existing problems and for them to track (i.e. check) that they've been dealt with before merging.  Using this tool as a part of our process would mean that it's not necessary for a reviewer to comment on conflicts, failed tests, or linting errors.  The author, by unchecking, has indicated that they are aware of the existing problems and that by unchecking, is acknowledging that they will be dealt with before merging.  If they intend to leave them not dealt with, they would have to comment on the unchecked item.

Each of the items I added in this PR template has a reason for being there.  I'm not adding them just for the sake of adding them.  They all serve a functional purpose.  One of those purposes is making sure we're all on the same page with our review processes, which has been a problem.  People have been operating in seemingly differing and opaque ways, leaving PR authors and/or reviewers waiting for things that aren't going to otherwise happen.

**Comment by hepcat72** [comment]
It would be nice to have the updated templates for the new round of Tracebase work...

**Comment by lparsons** [comment]
> It would be nice to have the updated templates for the new round of Tracebase work...

@hepcat72 Could you write up a short document explaining the process and how you intend these templates to be used? That will be useful to jumping off point for a discussion at our next developer meeting. Once we settle on things, we can get it added as a "standard operating procedure" in NPLC admin docs.

**Comment by hepcat72** [comment]
> > It would be nice to have the updated templates for the new round of Tracebase work...
> 
> @hepcat72 Could you write up a short document explaining the process and how you intend these templates to be used? That will be useful to jumping off point for a discussion at our next developer meeting. Once we settle on things, we can get it added as a "standard operating procedure" in NPLC admin docs.

I assume you're referring to the PR template.  The changes to the issue templates were mainly aesthetic. Suggestions were put in comments as opposed to in-line so that incomplete sections in submitted issues say "None provided" instead of the instructions. And the items in the developer section were just rearranged based on expected content size.  Smaller sections were moved to the top to be easier to find when requirements and design have numerous lines.

I will document what we previously discussed about the PR template.

**Comment by hepcat72** [comment]
> OK, let's get these implemented!

I had started working on the document you requested about a week ago and got distracted with the ATACC issues, but reading between the lines of your above comment and the discussion in the meeting, I assume you're referring to implementing the deadline suggestion and that the discussion takes the place of the doc.

- [ ] TODO: Add a deadline item to the template.

If you're referring to something else, could you elaborate? Or do you just mean "merge"?

**Comment by lparsons** [comment]
If you can get a document together, that would be very helpful. Just something short and sweet that describes how we'd use these various checkboxes, etc. If you want to include a deadline field, that's fine by me, it could be useful in some circumstances to keep things moving.

**Comment by hepcat72** [comment]
Created a doc to explain the changes in this PR...

[pr_template_update_requirements.md](https://github.com/Princeton-LSI-ResearchComputing/tracebase/files/12337721/pr_template_update_requirements.md)

**Comment by hepcat72** [comment]
@lparsons - I'm still waiting on linting, but I thought I should call your attention to the footnote I added about the review period, in which I took some liberties to prescribe a path to resolution if the review period lapses.  I included what was discussed in the meeting (reach out to reviewers and bring up in a meeting), but I also added a path to resolution in the event there is another review period lapse and limitations on putting a PR "on hold". Feel free to suggest changes or removal of the footnote. I'm just endeavoring to provide transparency and clarity in resolving hung PRs.

File: .github/pull_request_template.md [hunk 9]
```diff
@@ -2,49 +2,75 @@
...
+
+
+^ *Reviews must be completed within the review period (in work time) from the*
+*time of the review request.  If the (re-)review period lapses, the author*
+*must reach out to reviewers or request merge in a meeting.  Requests for more*
+*time must have a deadline (defaulting to the review period).  If the review*
+*period lapses again, as long as blocking issues are responded to, all other*
+*merge requirements are met, and the pull request has not been closed/rejected*
+*by any reviewer, the author may merge the PR.  PRs may not be put on hold for*
+*longer than the review period.*
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply .github/pull_request_template.md:76]: I can see how you might think this is about #569, but honestly, while the thought did occur to me that the mechanisms hammered out here could possibly be applied to that PR, this isn't about a specific PR. In fact, the PRs I was thinking about when making these template edits were the ones about the validation database removal (and various other previous PRs). WRT #569, I still intend to present my other presentation when we have time in the meetings before trying to close that loop.

My only purpose here is to define a process that leads to a decisive and timely conclusion. So I'm just trying to patch the gaps in the process that lead to a sort of "PR limbo", if you will.  Take your suggested edit for instance.  It's open ended in that it doesn't define how the project manager comes to face the decision to merge, or a timeline on when to merge, or what criteria are used in that decision, or what happens when the project manager does not wish to merge.  That is all (to use the transparency theme I was advocating for) rather opaque.  I think that it's entirely appropriate to make it the project manager's decision, and I thought about that as an option, but since the onus comes from the author, I thought that a solution that keeps the onus with the author would make it the most convenient solution - but to balance that "merge power", I decided to include language that states that any reviewer (which assumes the project manager to be among them) could close/reject the PR.  In that way, it's decisive.  It's a conclusion and it's timely.

I'm willing to go with your edit (or even deletion of the entire footnote if you prefer) to merge this PR as a step forward, but let me know what you think.

[the remark this answers is not in the document]
Comment (hepcat72) [reply .github/pull_request_template.md]: I couldn't decide on an edit that fixes the problem it was intended to fix, so I just removed it. Besides, I agree it's more of an internal thing.

**Review by lparsons — APPROVED** [review]
Looks good to me, thanks for the work on this @hepcat72.
